### PR TITLE
Ejector

### PR DESCRIPTION
**Pull request by jianoaix**
## Why are these changes needed?
Make an Ejector API for automated ejection of nonsigners based on objective onchain nonsigning rates and protocol SLA: https://docs.eigenlayer.xyz/eigenda/operator-guides/requirements/protocol-SLA.

## Checks

- [x] I've made sure the lint is passing in this PR.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, in that case, please comment that they are not relevant.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Integration tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: disperser/dataapi/ejector.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+func newEjector(logger logging.Logger, tx core.Transactor, metrics *Metrics) *ejector {
+	return &ejector{
+		logger:     logger.With("component", "Ejector"),
+		transactor: tx,
+		metrics:    metrics,
+	}
+}
+
+func (e *ejector) eject(ctx context.Context, nonsigningRate *OperatorsNonsigningPercentage, mode string) error {
```
Comment (ian-shim) [inline disperser/dataapi/ejector.go:56]: `mode` isn't used

File: disperser/dataapi/ejector.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
+package dataapi
```
Comment (ian-shim) [inline disperser/dataapi/ejector.go:1]: Let's add a test for ejector

File: disperser/dataapi/server.go [hunk 3]
```diff
@@ -292,6 +301,68 @@ func (s *server) Shutdown() error {
...
+//	@Produce	json
+//	@Param		interval query int false "Lookback window for operator ejection [default: 86400]"
+//	@Param		end  query int false "End time for evaluating operator ejection [default: now]"
+//	@Param		mode query string "Whether it's periodic or urgent ejection request [default: periodic]"
+//	@Success	200			{object}	BlobMetadataResponse
+//	@Failure	400			{object}	ErrorResponse	"error: Bad request"
+//	@Failure	404			{object}	ErrorResponse	"error: Not found"
+//	@Failure	500			{object}	ErrorResponse	"error: Server error"
+//	@Router		/ejector/ejection [get]
+func (s *server) EjectOperatorsHandler(c *gin.Context) {
```
Comment (ian-shim) [inline disperser/dataapi/server.go:317]: Let's add a test on `server_test`

**Review by ian-shim — COMMENTED** [review]
👏👏👏

One general question. It's fine for this PR as we want this functionality asap, but I'm not sure if `dataapi` is the right place to host this method. `dataapi` is meant to serve (readonly) data, and we eventually want to make it available for external parties to run as well.

File: disperser/dataapi/ejector.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+// operatorPerfScore scores an operator based on its stake share and nonsigning rate. The
+// performance score will be in range [0, 1], with higher score indicating better performance.
+func operatorPerfScore(stakeShare float64, nonsigningRate float64) float64 {
+	if nonsigningRate == 0 {
+		return 1.0
+	}
+	sla := stakeShareToSLA(stakeShare)
+	perf := (1 - sla) / nonsigningRate
+	return perf / (1.0 + perf)
```
Comment (mooselumph) [inline disperser/dataapi/ejector.go:32]: I'm a bit uncertain about whether this scoring will give us the right ordering across SLA classes. Interested to know if there was a rationale for this particular scoring. I'll put some further thought into it.

**Comment by jianoaix** [comment]
> 👏👏👏
> 
> One general question. It's fine for this PR as we want this functionality asap, but I'm not sure if `dataapi` is the right place to host this method. `dataapi` is meant to serve (readonly) data, and we eventually want to make it available for external parties to run as well.

a control plane restructuring will be needed later. For now this will be permissioned at upper layer as well as inside the server (with the secrete token)

**Review by mooselumph — APPROVED** [review]

File: disperser/dataapi/ejector.go (hunk 1 above)
Comment (jianoaix) [reply disperser/dataapi/ejector.go]: Actually this isn't needed even after fixing TODO -- removed it!

**Pull request merged**